Added validation for pods via checks on attached replica sets

Danil-Grigorev opened the pull request:
Adding health checks for the app as a part of migplan.
New status field:
```
 kind: MigMigration
...
  status:
    conditions:
    - category: Warn
      lastTransitionTime: 2019-10-14T10:36:59Z
      message: The destination cluster has unhealthy migrated namespaces [nginx-example].
        See status.namespaces for details.
      status: "True"
      type: UnhealthyDestinationNamespaces
    - category: Advisory
      durable: true
      lastTransitionTime: 2019-10-14T10:36:59Z
      message: The migration has completed successfully.
      reason: Completed
      status: "True"
      type: Succeeded
    namespaces:
    - name: nginx-example
      workloads:
      - name: Deployment/nginx-deployment
        resources:
        - Pod/nginx-deployment-9b96654cd-l27kh
    phase: Completed
    startTimestamp: 2019-10-14T10:35:28Z
```

jortel commented on pkg/controller/migmigration/postValidate.go:
```diff
@@ -0,0 +1,298 @@
+		return false, err
+	}
+
+	if failedPodCounter > 0 {
```
Looks like only the _category_ is different between the `if` and `else`.  Perhaps it would be good to reduce the duplication by:
```
category := Advisory
if failedPodCounter > 0 {
    category = Critical
}
```

Then:  `Category: category`.

jortel replied:
Thinking this should not be a `Critical` category but rather see `Reason: Failed` below.

jortel commented on pkg/controller/migmigration/postValidate.go:
```diff
@@ -0,0 +1,298 @@
+)
+
+// EventDisplay is a structure for displaying occured events
+type EventDisplay struct {
```
Why not just use the `core.Event`.  Also, the `Display` in the name describes a specific usage of the data instead of being descriptive of the data itself.  Suggest dropping the _display_ terminology.

jortel commented on pkg/controller/migmigration/postValidate.go:
```diff
@@ -0,0 +1,298 @@
+
+const (
+	// ValidationFailed is a failed validation state
+	ValidationFailed = iota
```
The term `Validation` has a pretty specific meaning in our controller/CRs.  I pertains to assessing the stage of the world prior to running a migration.  Suggest something like `Verification` for the _post_ checks would be clearer.

That said, I think you can replace these return values by having the task code chec the `Reason = (Running|Succeeded|Failed)` in the `PostVerification` (or some other name) condition.  Will describe condition changes elsewhere.

Danil-Grigorev replied:
Thanks for pointing it out. Going to replace the naming, sorry for confusion.

jortel commented on pkg/controller/migmigration/postValidate.go:
```diff
@@ -0,0 +1,298 @@
+	filteredEventMessages := []EventDisplay{}
+
+	for _, event := range eventList {
+		if !involvedObjectMatch(event, objectNames) || event.Type != corev1.EventTypeWarning {
```
Put the `event.Type != corev1.EventTypeWarning` first in the `||` for efficiency.

jortel commented on pkg/controller/migmigration/postValidate.go:
```diff
@@ -0,0 +1,298 @@
+	}
+
+	t.Owner.Status.SetCondition(migapi.Condition{
+		Type:     ValidationDetails,
```
A separate `details` condition is a bit odd.  Let's consider a single condition.  Our conditions have an `Items []string` field that is handy for this.

Thinking something like this:

```
t.Owner.Status.SetCondition(migapi.Condition{
    Type:     Verification,
    Status:   True,
    Reason:   Running,
    Category: Advisory,
    Message:  "Verification running."
    Items: []string{}
})
```
Then when a Pod fails verification.
```
condition := t.Owner.Status.FindCondition(Verification)
for _, event := range events {
    condition.Items = append(condition.Items, string(event-json))
}
```

At the end,
```
condition := t.Owner.Status.FindCondition(Verification)
if len(condition.Items) > 0 {
    condition.Category = Critical
    condition.Reason = Failed
    condition.Message = "Verification failed: []"  // The [] is replaced Items by when staging ends.
} else {
    condition.Reason = Succeeded
    condition.Message = "Verification succeeded." 
}
```

jortel replied:
As we discussed: Rather than listing events, let's provide the user with a list of resources that failed (Deployment|ReplicaSet, ..)
```
condition := t.Owner.Status.FindCondition(Verification)
for _, r := range resources {
    ref := fmt.Sprintf("%s/%s:%s", r.Kind, r.Namespace, r.Name)
    condition.Items = append(condition.Items, ref)
}
```

jortel commented on pkg/controller/migmigration/task.go:
```diff
@@ -41,6 +41,9 @@ const (
 	FinalRestoreCreated           = "FinalRestoreCreated"
 	FinalRestoreFailed            = "FinalRestoreFailed"
+	ValidateResourcesMigrated     = "ValidateResourcesMigrated"
+	ValidationInProgress          = "ValidationInProgress"
```
s/Validation/Verification/g

A separate `InProgress` phase deviates from other phases that can also be long running.  Also, I believe that the `ValidateResourcesMigrated` phase checks the static resources like DCs.  Suggest something like:

VerifyPodOwners
VerifyPods

.. or something :)

jortel commented on pkg/controller/migmigration/validation.go:
```diff
@@ -15,6 +15,8 @@ const (
 	Succeeded         = "Succeeded"
 	Failed            = "Failed"
+	ValidationDetails = "ValidationDetails"
+	Validating        = "Validating"
```
Suggest (1) phase named: `Verification`.

Danil-Grigorev replied:
Updated.

jortel commented on pkg/controller/migmigration/task.go:
```diff
@@ -465,8 +470,32 @@ func (t *Task) Run() error {
+	case InitialBackupFailed, FinalRestoreFailed, ResourceValidationFailed:
 		t.next()
+	case ValidateResourcesMigrated:
+		migratedSuccessfully, err := t.validateStaticResourcesPresence()
```
I'm not convinced this adds value.  It is essentially testing that velero backed up and restored all of the resources expected. Should this fail for one migration, our controller is broken and it will fail for all of them.  Seems this should be caught in e2e tests.

Danil-Grigorev replied:
Removed after short discussion. This check will make migration fail every time, and it should be a part of pre-migration health checks. @alexander-demichev We probably need to check if our migration tooling tag all the resources in the namespace, so that velero could pick it up. For example `ReplicaSet` created by an older version of the `Deployment` is left in the namespace, but is not picked by our migration. @jwmatthews Is this expected behaviour?

jortel replied:
You'll need to consider the owner references.  For example: for a Deployment <- ReplicaSet <- Pod  only the `Deployment` should be backed/Restored by velero since the Deployment will _spin up_ the ReplicaSet which will, in turn, _spin up_ the Pod.

dymurray commented on pkg/controller/migmigration/verify.go:
```diff
@@ -0,0 +1,329 @@
+			if err != nil {
+				return VerificationFailed, err
+			}
+			// Skip resource with uprepared status
```
NIT: `unprepared`

dymurray commented on pkg/controller/migmigration/verify.go:
```diff
@@ -0,0 +1,329 @@
+			}
+			// Skip resource with uprepared status
+			if !found {
+				_, found, err := unstructured.NestedFieldCopy(podManager.Object, "status", "unavailableReplicas")
```
This block feels weird... So if the `readyReplicas` field isn't found, we then check for `unavailableReplicas`. If `unavailableReplicas` aren't found, we return an error that `replicas field was not found in kind`.

Do we need two conditionals here? Won't `err` be set if `NestedFieldCopy` fails when `unavailableReplicas` doesn't exist? If that's the case we can remove the `if !found` block on L140. If the error message is nonsensical then I can understand having the two conditions here.

If `unavailableReplicas` are found I'm assuming we are continuing to get the failing pods health?

Danil-Grigorev replied:
After some investigation, I removed this part. Just looking into the `Pods` is enough. The error suggests only that there was some non string accessor in the structure, so it does not give anything. And another thing - the `unavailableReplicas` is also for pods in `Pending` state :(

dymurray commented on pkg/controller/migmigration/verify.go:
```diff
@@ -0,0 +1,329 @@
+		// Consider suspicious containers as an error state
+		podContainersStatuses := append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...)
+		for _, containerStatus := range podContainersStatuses {
+			if !containerStatus.Ready && containerStatus.RestartCount >= CrashLoopBackOffMargin {
```
I'm okay with this. We are saying ONLY if the pod is not ready will we do a check to see if there are more than 3 restarts.

dymurray commented:
In my opinion it sounds like the preferred method here would be #2. I think closing a plan should prevent subsequent migrations from occurring, but validating the state of that migration plan should occur after the plan is closed.

My only concern with that is that we would be running validation indefinitely on the migration plan until it is deleted. I'm not quite sure that is desired behavior considering that the controller will be allocating resources to a plan that could be days old and the state of the application wouldn't have changed. Perhaps once the validation succeeds completely then we set another field such as `appVerified` which prevents the verification loop from running.

The more I write I realize I'm basically talking about option 1, but I think it's important that the terminology remains the same. I don't think app verification should affect the existing workflow with how `closed` is being used today from the UI. So that is my hard requirement with a solution here. Curious what @jortel thinks.

jortel commented on pkg/apis/migration/v1alpha1/verify.go:
```diff
@@ -0,0 +1,208 @@
+
+// UnhealthyResources is a store for unhealthy namespaces across clusters
+type UnhealthyResources struct {
+	SourceUnhealthy      []UnhealthyNamespace `json:"sourceUnhealthy,omitempty"`
```
I think that simply `Souce` and `Destination` field names would be clearer.

dymurray commented on pkg/apis/migration/v1alpha1/verify.go:
```diff
@@ -0,0 +1,208 @@
+// AddUnhealthyResources is adding unhealthy workloads and resolves owner references
+func (u *UnhealthyResources) AddUnhealthyResources(
+	client k8sclient.Client,
+	targetCluster *[]UnhealthyNamespace,
```
Consider a better name for this. Even `unhealthyNamespaces` would suffice. `targetCluster` makes me think the object is a client or something.

jortel commented:
I don't think there should be any further processing on a closed plan.  I also don't think we want the verification to run on every plan reconcile.  Instead, it should only run after a successful _final_ migration.  The question is - how should the verification get triggered?  Let's discuss.

Danil-Grigorev commented:
> Per the discussion on the email thread, I believe this PR is still missing what we agreed upon for resolving this with plan closure. The `verify` boolean on the plan spec is supposed to be used by the migration controller to determine whether or not we will run application verification on the destination cluster at the end of migration.

Changed.
> Currently this PR is validating the application state of the source and destination everytime the plan controller reconciles. You can leave the source validation code on the plan controller, but the destination cluster verification needs to be done on the migration controller as a phase at the end of migration.

I removed the part of the code responsible for source health checks. The health should be shown on the migPlan status. Should be added later by @alexander-demichev, me or anyone else.

jortel commented on pkg/apis/migration/v1alpha1/resource.go:
```diff
@@ -6,6 +6,7 @@ import (
 
 const (
 	TouchAnnotation   = "touch"
+	VerifyAnnotation  = "verify"
```
The annotation is no longer needed.

jortel commented on pkg/apis/migration/v1alpha1/resource.go:
```diff
@@ -70,6 +71,26 @@ func (r *MigPlan) Touch() {
 	r.Annotations[TouchAnnotation] = uuid.New().String()
 }
 
+func (r *MigPlan) GetVerify() (verify bool) {
```
The annotation is no longer needed.  Just a `bool`.

Danil-Grigorev replied:
Sorry, this part was reintroduced after rebase, will clean it up.

jortel commented on pkg/apis/migration/v1alpha1/verify.go:
```diff
@@ -0,0 +1,248 @@
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Types
```
To keep condition `const` in (1) place, we have been defining them in `validation.go` even though this is verification.  Please put them there.

jortel commented on pkg/apis/migration/v1alpha1/verify.go:
```diff
@@ -0,0 +1,248 @@
+
+// Types
+const (
+	UnhealthySourceNamespaces      = "UnhealthySourceNamespaces"
```
The migration controller verification only deals with the destination.  No need for Source/Destination differentiation or for both.  Just `UnhealthyNamespaces`.

jortel commented on pkg/apis/migration/v1alpha1/verify.go:
```diff
@@ -0,0 +1,248 @@
+
+// Messages
+const (
+	UnhealthyDestination = "The destination cluster has unhealthy migrated namespaces []. See status.namespaces for details."
```
Since the `UnhealthyNamespaces` stucture contains the full list/description, I think the message should contain the list `[]` as well.

Danil-Grigorev replied:
From what I see in the message, there is a list `[]` already. Could you please clarify?

jortel replied:
Meant should **not** include the `[] ` list.

jortel commented on pkg/apis/migration/v1alpha1/verify.go:
```diff
@@ -0,0 +1,248 @@
+}
+
+// VerifyPods would collect, display and notify about the pod health in a specific namespace
+func (u *UnhealthyResources) VerifyPods(client k8sclient.Client, options *k8sclient.ListOptions) (*[]unstructured.Unstructured, error) {
```
`VerifyPods()` seems like a method on the Reconciler not the `UnhealthyResources` object.

Danil-Grigorev replied:
So, this mean that until the `spec.verify` is set on the migPlan, the migrations will verify target namespaces on every reconcile?

Danil-Grigorev replied:
I suppose this part could be used at any moment, and not only by migration/plan reconsile.

jortel replied:
No, the migration controller will only do the verification when verify=true and the migration is a _final_ migration that has completed successfully.

jortel commented on pkg/apis/migration/v1alpha1/verify.go:
```diff
@@ -0,0 +1,248 @@
+
+	return false
+}
+
```
I would expect these _find_ methods to be public and used by the controller, reconciler.

Danil-Grigorev replied:
Sure, could be fixed.

jortel commented on pkg/apis/migration/v1alpha1/verify.go:
```diff
@@ -0,0 +1,248 @@
+}
+
+// AddUnhealthyNamespace is adding a workload to selected list of unhealthy namespaces
+func (u *UnhealthyResources) AddUnhealthyNamespace(cluster *[]UnhealthyNamespace, namespace, name string, resourceNames ...string) {
```
This method is scoped to the `UnhealthyResources` object.  Anything more than `AddNamespace()` is redundant.

jortel commented on pkg/apis/migration/v1alpha1/verify.go:
```diff
@@ -0,0 +1,248 @@
+}
+
+// AddUnhealthyResources is adding unhealthy workloads and resolves owner references
+func (u *UnhealthyResources) AddUnhealthyResources(
```
This method is scoped to the `UnhealthyResources` object.  Anything more than `AddResources()` is redundant.

jortel commented on pkg/controller/migmigration/task.go:
```diff
@@ -41,6 +41,7 @@ const (
 	EnsureFinalRestore            = "EnsureFinalRestore"
 	FinalRestoreCreated           = "FinalRestoreCreated"
 	FinalRestoreFailed            = "FinalRestoreFailed"
+	Verify                        = "Verify"
```
The name `Verification` seems more like phase name.  Eg.  _"The migration is in the verification phase."_

Danil-Grigorev replied:
Updated as well.

Danil-Grigorev commented:
Some structure improvements are needed, as the feature complexity is higher then expected.

jortel commented on pkg/controller/migmigration/validation.go:
```diff
@@ -46,6 +48,8 @@ const (
 	FailedMessage            = "The migration has failed.  See: Errors."
 	SucceededMessage         = "The migration has completed successfully."
+	UnhealthyDestination     = "The destination cluster has unhealthy migrated namespaces. See status.namespaces for details."
+	UnhealthySource          = "The source cluster has unhealthy namespaces. See status.namespaces for details."
```
The `UnhealthySource` would be used by the plan controller and should be defined there.  Also, by convention, the _message_ const should end with `Message`.  Recommend: `UnhealthyNamespacesMessage` to match `UnhealthyNamespaces` condition type.
